sdks/java/core/src/main/java/org/apache/beam/sdk/options/ValueProvider.java

```diff
@@ -0,0 +1,226 @@
+  /**
+   * Return the value wrapped by this {@link ValueProvider}.
+   */
+  T get();
+
+  /**
+   * Whether the contents of this {@link ValueProvider} is available to validation
+   * routines that run at graph construction time.
+   */
+  boolean shouldValidate();
+
+  /**
+   * {@link StaticValueProvider} is an implementation of {@link ValueProvider} that
+   * allows for a static value to be provided.
+   */
+  @JsonSerialize(using = ValueProvider.Serializer.class)
+  @JsonDeserialize(using = ValueProvider.Deserializer.class)
+  public static class StaticValueProvider<T> implements ValueProvider<T>, Serializable {
+    @Nullable
+    private final T value;
+
+    StaticValueProvider(@Nullable T value) {
+      this.value = value;
+    }
+
+    /**
+     * Creates a {@link StaticValueProvider} that wraps the provided value.
+     */
+    public static <T> StaticValueProvider<T> of(T value) {
+      StaticValueProvider<T> factory = new StaticValueProvider<>(value);
+      return factory;
+    }
+
+    @Override
+    public T get() {
+      return value;
+    }
+
+    @Override
+    public boolean shouldValidate() {
+      return true;
+    }
+  }
+
+  /**
+   * {@link RuntimeValueProvider} is an implementation of {@link ValueProvider} that
+   * allows for a value to be provided at execution time rather than at graph
+   * construction time.
+   *
+   * <p>To enforce this contract, if there is no default, users must only call
+   * {@link #get()} at execution time (after a call to {@link Pipeline#run}),
+   * which will provide the value of OPTIONS.
+   */
+  @JsonSerialize(using = ValueProvider.Serializer.class)
+  @JsonDeserialize(using = ValueProvider.Deserializer.class)
+  public static class RuntimeValueProvider<T> implements ValueProvider<T>, Serializable {
+    private static ConcurrentHashMap<Long, PipelineOptions> optionsMap =
+      new ConcurrentHashMap<>();
+
+    private final Class<? extends PipelineOptions> klass;
+    private final String methodName;
+    private final T defaultValue;
+    private final Long optionsId;
+
+    /**
+     * Creates a {@link RuntimeValueProvider} that will query the provided
+     * optionsId for a value.
+     */
+    RuntimeValueProvider(String methodName, Class<? extends PipelineOptions> klass,
+                         Long optionsId) {
+      this.methodName = methodName;
+      this.klass = klass;
+      this.defaultValue = null;
+      this.optionsId = optionsId;
+    }
+
+    /**
+     * Creates a {@link RuntimeValueProvider} that will query the provided
+     * optionsId for a value, or use the default if no value is available.
+     */
+    RuntimeValueProvider(String methodName, Class<? extends PipelineOptions> klass,
+      T defaultValue, Long optionsId) {
+      this.methodName = methodName;
+      this.klass = klass;
+      this.defaultValue = defaultValue;
+      this.optionsId = optionsId;
+    }
+
+    static void setRuntimeOptions(PipelineOptions runtimeOptions) {
```

Comment:
comment for readers. Maybe something like `Once set, all {@code RuntimeValueProviders} will return {@code true} from {@code isAccessible()}. By default, the value is set when deserializing {@link PipelineOptions}.`